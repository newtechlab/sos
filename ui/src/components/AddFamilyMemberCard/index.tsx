import styled from "styled-components";
import { StyledCard } from "../StyledFamilyCard";

interface AddFamilyMemberCardProps {
    onClick: () => void;
}

export default function AddFamilyMemberCard(props: AddFamilyMemberCardProps) {
    const { onClick } = props;

    return <StyledCard onClick={() => onClick()}>
        <CenterTextDiv>+ Legg til</CenterTextDiv>
    </StyledCard>
}

const CenterTextDiv = styled.div`
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
`
