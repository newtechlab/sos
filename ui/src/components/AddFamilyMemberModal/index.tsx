import { Button, Card, Dropdown, Header, Icon, Image, Input, Modal } from "semantic-ui-react";
import { FamilyMember } from "../../App";
import family from './family.png';
import { v4 as uuidv4 } from 'uuid';

interface AddFamilyMemberModalProps {
    open: boolean;
    setOpen: (_: boolean) => void; 
    addFamilyMember: (_: FamilyMember) => void;
}

const ageOptions = [
    {
      key: 'baby',
      text: 'baby',
      value: 'baby',
    //   image: { avatar: true, src: '/images/avatar/small/jenny.jpg' },
    },
    {
      key: 'adult',
      text: 'adult',
      value: 'aduly',
    //   image: { avatar: true, src: '/images/avatar/small/elliot.jpg' },
    }
  ]

export default function AddFamilyMemberModal(props: AddFamilyMemberModalProps) {
    const { open, setOpen, addFamilyMember } = props;

    const newFamilyMember: FamilyMember = {
        id: uuidv4(), 
        age: 0
      }

    return <Modal
        size="small"
        onClose={() => setOpen(false)}
        onOpen={() => setOpen(true)}
        open={open}
        trigger={<Button>Show Modal</Button>}
    >
        <Modal.Header>Legg til familiemedlem</Modal.Header>
        <Modal.Content image>      
            <Image size='medium' src={family} wrapped />
        </Modal.Content>
        <Modal.Actions>
            <Input placeholder="Bob" />
            <Dropdown
                placeholder='Age'
                selection
                options={ageOptions}
            />
            <Button color='black' onClick={() => setOpen(false)}>
                cancel
            </Button>
            <Button
                content="Add"
                positive
                onClick={() => {
                    addFamilyMember(newFamilyMember)
                    setOpen(false)
                }}
            />
        </Modal.Actions>
    </Modal>
}