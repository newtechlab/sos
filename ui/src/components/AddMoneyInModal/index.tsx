import { Button, Container, Dropdown, Grid, Input, Modal } from "semantic-ui-react";
import { LedgerRow, StyledOverridesDiv, TransactionCategory } from "../../App";
import { v4 as uuidv4 } from "uuid";
import { useEffect, useState } from "react";
import DaySelector from "../DaySelector";
import styled from "styled-components";

interface AddMoneyInModalProps {
  open: boolean;
  setOpen: (_: boolean) => void;
  addLedgerRow: (_: LedgerRow) => void;
}

interface MoneyInAndCategory {
  name: string;
  category: TransactionCategory;
}

interface DropDownItem {
  key: string;
  text: string;
  value: string;
}

const convertDropdownItem = (item: MoneyInAndCategory): DropDownItem => {
  return {
    key: item.name,
    text: item.name,
    value: item.name,
  };
};

export default function AddMoneyInModal(props: AddMoneyInModalProps) {
  const [from, setFrom] = useState<string | undefined>(undefined);
  const [moneyInItems, setMoneyInItems] = useState<
    Map<string, MoneyInAndCategory>
  >(new Map<string, MoneyInAndCategory>());  
  const [dropDownItems, setDropDownItems] = useState<DropDownItem[]>([]);
  const [amount, setAmount] = useState<number | undefined>(undefined);
  const [day, setDay] = useState<number | undefined>(undefined);
  const { open, setOpen, addLedgerRow } = props;

  useEffect(() => {
    const items = new Map<string, MoneyInAndCategory>();
    items.set("Lönn", {
      name: "Lönn",
      category: TransactionCategory.Income,
    });
    items.set("Pension", {
      name: "Pension",
      category: TransactionCategory.Income,
    });
    items.set("Benefit", {
      name: "Benefit",
      category: TransactionCategory.Income,
    });       
    setMoneyInItems(items);
  }, []);

  useEffect(() => {
    const dropDownItems: DropDownItem[] = [];
    moneyInItems.forEach((value) => {
      const i = convertDropdownItem(value);
      dropDownItems.push(i);
    });
    setDropDownItems(dropDownItems);
  }, [moneyInItems]);  

  return (
    <Modal
      onClose={() => setOpen(false)}
      onOpen={() => setOpen(true)}
      open={open}
      trigger={<Button>Show Modal</Button>}
    >
      <StyledModalContent>
        <StyledOverridesDiv>
          <Container>
            <StyledContainerSpace>
              <h1> Ny inntekt </h1>

              <StyledModalBody>
                <StyledIngress>
                  Penger du får inn på konto jevnlig. Det kan være lønn,
                  barnebidrag, trygd eller annen støtte. Det man ikke bør legge
                  inn er gaver eller lignende, fordi man ikke kan regne med at
                  det kommer inn jevnlig og derfor ikke bør regnes med i et
                  budsjett.
                </StyledIngress>
                <Grid columns={2}>
                  <Grid.Column width={2}>Type</Grid.Column>
                  <Grid.Column width={14}>
                    <Dropdown
                      placeholder="f.eks. Lønn"
                      options={dropDownItems}
                      onChange={(_, data) => {
                        setFrom(data.value?.toString());
                      }}
                    />
                  </Grid.Column>
                  <Grid.Column width={2}>Beløp</Grid.Column>
                  <Grid.Column width={14}>
                    <Input
                      placeholder="f.eks 10 000"
                      onChange={(_, data) => {
                        setAmount(parseInt(data.value?.toString() || "0", 10));
                      }}
                    />
                  </Grid.Column>
                </Grid>
                <br />
                <DaySelector
                  itemSelected={(day) => {
                    setDay(day);
                  }}
                />
                <br />
                <Grid columns={2}>
                  <Grid.Column>
                    <Button
                      circular
                      basic
                      color="blue"
                      onClick={() => setOpen(false)}
                    >
                      Avbryt
                    </Button>
                  </Grid.Column>
                  <Grid.Column textAlign="right">
                    <Button
                      content="Legg til"
                      color="blue"
                      onClick={() => {
                        if (from && amount && day) {
                          addLedgerRow({
                            id: uuidv4(),
                            dayOfMonth: day,
                            amount: amount,
                            accountFrom: from,
                            accountTo: "user",
                            category: TransactionCategory.Income,
                          });
                          setOpen(false);
                        } else {
                          console.log("missing var", from);
                          console.log("missing var", amount);
                          console.log("missing var", day);
                        }
                      }}
                    />
                  </Grid.Column>
                </Grid>
              </StyledModalBody>
            </StyledContainerSpace>
          </Container>
        </StyledOverridesDiv>
      </StyledModalContent>

      {/* <Modal.Actions> */}

      {/* <Dropdown
                placeholder='Day of Month'
                selection
                options={getDaysOfMonthDropdown()}
                onChange={ (_, data) => { 
                    setDay(parseInt(data.value?.toString() || "1", 10)) 
                }}
            /> */}

      {/* </Modal.Actions> */}
    </Modal>
  );
}

export const StyledModalContent = styled(Modal.Content)`
  padding: 0 !important;
`;

export const StyledModalBody = styled.div`
  background-color: #ffffff;
  padding: 20px;
`;

const StyledContainerSpace = styled.div`
  padding: 3em;
`;
const StyledIngress = styled.div`
  padding-bottom: 2em;
`;
