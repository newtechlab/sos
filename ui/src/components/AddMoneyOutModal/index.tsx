import {
  Button,
  Dropdown,
  Grid,
  GridColumn,
  GridRow,
  Icon,
  Input,
  Modal,
} from "semantic-ui-react";
import { LedgerRow, TransactionCategory } from "../../App";
import { v4 as uuidv4 } from "uuid";
import { useEffect, useState } from "react";
import { StyledGrid, StyledGridRowBottom } from "../MoneyIn";
import ErrorBar from "../ErrorBar";

let belopError = false;
const validateInput = (inputvalue: string) => {
  const validNumber = RegExp(/^[0-9\b]+$/);
  validNumber.test(inputvalue) ? (belopError = false) : (belopError = true);
};
interface AddMoneyOutModalProps {
  open: boolean;
  setOpen: (_: boolean) => void;
  addLedgerRow: (_: LedgerRow) => void;
  categories: Set<string>;
  header: string;
  ingresstext: string;
}

interface MoneyOutAndCategory {
  name: string;
  category: TransactionCategory;
}

interface DropDownItem {
  key: string;
  text: string;
  value: string;
}

const convertDropdownItem = (item: MoneyOutAndCategory): DropDownItem => {
  return {
    key: item.name,
    text: item.name,
    value: item.name,
  };
};

export default function AddMoneyOutModal(props: AddMoneyOutModalProps) {
  const [to, setTo] = useState<string | undefined>(undefined);
  const [moneyOutItems, setMoneyOutItems] = useState<
    Map<string, MoneyOutAndCategory>
  >(new Map<string, MoneyOutAndCategory>());
  const [dropDownItems, setDropDownItems] = useState<DropDownItem[]>([]);
  const [amount, setAmount] = useState<number | undefined>(undefined);
  const [day] = useState<number | undefined>(1);
  const { open, setOpen, addLedgerRow } = props;

  useEffect(() => {
    const items = new Map<string, MoneyOutAndCategory>();
    items.set("Abonnement (Netflix, HBO, Spill m.m)", {
      name: "Abonnement (Netflix, HBO, Spill m.m)",
      category: TransactionCategory.Media_and_Subscriptions,
    });
    items.set("Sport, trening og fritid", {
      name: "Sport, trening og fritid",
      category: TransactionCategory.Recreation,
    });
    items.set("Transport (trikk, tog, buss)", {
      name: "Transport (trikk, tog, buss)",
      category: TransactionCategory.Travel_Expenses,
    });
    items.set("Huslån", {
      name: "Huslån",
      category: TransactionCategory.Debt,
    });
    items.set("Husleie", {
      name: "Husleie",
      category: TransactionCategory.Debt,
    });
    items.set("Reise (Fly, hotell)", {
      name: "Reise (Fly, hotell)",
      category: TransactionCategory.Travel_Expenses,
    });
    items.set("Mat og drikke", {
      name: "Mat og drikke",
      category: TransactionCategory.Food_and_Beverages,
    });
    items.set("Strøm", {
      name: "Strøm",
      category: TransactionCategory.Electricity,
    });
    items.set("Klær og sko", {
      name: "Klær og sko",
      category: TransactionCategory.Clothing_and_Footwear,
    });
    items.set("Personlig pleie", {
      name: "Personlig pleie",
      category: TransactionCategory.Personal_Care,
    });
    items.set("Forsikring", {
      name: "Forsikring",
      category: TransactionCategory.Insurance,
    });
    items.set("Andre bokostnader (kommunale avgifter m.m.)", {
      name: "Andre bokostnader (kommunale avgifter m.m.)",
      category: TransactionCategory.Household_Items,
    });
    items.set("Husholdningsartikler", {
      name: "Husholdningsartikler",
      category: TransactionCategory.Household_Items,
    });
    items.set("Annen gjeld", {
      name: "Annen gjeld",
      category: TransactionCategory.Debt,
    });
    items.set("Billån", {
      name: "Billån",
      category: TransactionCategory.Debt,
    });
    items.set("Faste donasjoner", {
      name: "Faste donasjoner",
      category: TransactionCategory.Personal_Care,
    });
    setMoneyOutItems(items);
  }, []);

  useEffect(() => {
    const dropDownItems: DropDownItem[] = [];
    moneyOutItems.forEach((value) => {
      if (props.categories.has(value.category)) {
        const i = convertDropdownItem(value);
        dropDownItems.push(i);
      }
    });
    setDropDownItems(dropDownItems);
  }, [moneyOutItems]);

  return (
    <Modal
      onClose={() => setOpen(false)}
      onOpen={() => setOpen(true)}
      open={open}
      trigger={<Button>Show Modal</Button>}
    >
      <Modal.Header>Legg til ny {props.header}</Modal.Header>
      <Modal.Content>{props.ingresstext}</Modal.Content>
      <Modal.Actions>
        <StyledGrid>
          <StyledGridRowBottom>
            {/* <Input 
                placeholder="Coffee" 
                onChange={ (_, data) => { setTo(data.value?.toString()) }} /> */}
            <Grid.Column width={2}>
              <Button
                circular
                basic
                color="blue"
                onClick={() => setOpen(false)}
              >
                Avbryt
              </Button>
            </Grid.Column>
            <Grid.Column width={5}>
              <Dropdown
                placeholder="for eksempel Husleie"
                selection
                options={dropDownItems}
                onChange={(_, data) => {
                  setTo(data.value?.toString());
                }}
              />
            </Grid.Column>
            <Grid.Column width={5}>
              <Input
                placeholder="Beløp"
                onChange={(_, data) => {
                  setAmount(parseInt(data.value?.toString() || "0", 10));
                  validateInput(data.value);
                }}
              />
              {belopError ? (
                <ErrorBar msg="Vennligst skriv inn et nummer" />
              ) : (
                ""
              )}
            </Grid.Column>
            {/* <Dropdown
                placeholder='Day of Month'
                selection
                options={getDaysOfMonthDropdown()}
                onChange={ (_, data) => { 
                    setDay(parseInt(data.value?.toString() || "1", 10)) 
                }}
            /> */}
            <Grid.Column width={3}>
              <Button
                circular
                color="blue"
                onClick={() => {
                  if (to && amount && day) {
                    addLedgerRow({
                      id: uuidv4(),
                      dayOfMonth: day,
                      amount: amount,
                      accountFrom: "user",
                      accountTo: to,
                      category:
                        moneyOutItems.get(to)?.category ||
                        TransactionCategory.Undefined,
                    });
                    setOpen(false);
                  } else {
                    console.log("missing var", to);
                    console.log("missing var", amount);
                    console.log("missing var", day);
                  }
                }}
              >
                <Icon name="plus" />
                Legg til
              </Button>
            </Grid.Column>
          </StyledGridRowBottom>
        </StyledGrid>
      </Modal.Actions>
    </Modal>
  );
}
