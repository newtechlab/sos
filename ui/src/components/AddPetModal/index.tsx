import { Button, Image, Grid, Icon, Input, Modal } from "semantic-ui-react";
import { LedgerRow, Pet, TransactionCategory } from "../../App";
import { v4 as uuidv4 } from "uuid";
import { useEffect, useState } from "react";
import { StyledGrid, StyledGridRow, StyledGridRowBottom } from "../MoneyIn";
import { StyledCard } from "../StyledFamilyCard";
import styled from "styled-components";
import dog from "./dog.png";

interface AddPetProps {
  open: boolean;
  setOpen: (_: boolean) => void;
  pets: Array<Pet>;
  setPets: (_: Array<Pet>) => void;
}

export default function AddPetModal(props: AddPetProps) {
  const [pet, setPet] = useState<string | undefined>(undefined);
  const { open, setOpen, pets, setPets } = props;

  return (
    <Modal onClose={() => setOpen(false)} open={open} pets={pets}>
      <Modal.Header>Legg til dyr</Modal.Header>
      <StyledImageBackground>
        <Modal.Content image>
          <StyledImageSpace>
            <Image size="medium" src={dog} wrapped />
          </StyledImageSpace>
        </Modal.Content>
      </StyledImageBackground>
      {/* <Modal.Content>
        Dina dyr, tex hund eller katt
      </Modal.Content> */}
      <Modal.Actions>
        <StyledGrid>
          <StyledGridRowBottom>
            <Grid.Column width={3}>
              <Button
                circular
                basic
                color="blue"
                onClick={() => setOpen(false)}
              >
                Avbryt
              </Button>
            </Grid.Column>
            {/* <Input 
                placeholder="Coffee" 
                onChange={ (_, data) => { setTo(data.value?.toString()) }} /> */}
            <Grid.Column width={8}>
              <Input
                placeholder="Skriv inn dyret her (f.eks Hest)"
                style={{ width: "100%" }}
                onChange={(_, data) => {
                  setPet(data.value?.toString());
                }}
              />
            </Grid.Column>

            <Grid.Column width={4}>
              <Button
                circular
                color="blue"
                onClick={() => {
                  if (pet) {
                    setPets([...pets, { id: uuidv4(), name: pet, type: pet }]);
                  }
                  setOpen(false);
                }}
              >
                <Icon name="plus" />
                Legg til
              </Button>
            </Grid.Column>
          </StyledGridRowBottom>
        </StyledGrid>
      </Modal.Actions>
    </Modal>
  );
}

const CenterTextDiv = styled.div`
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
`;

const AddPetCard = styled.div`
  margin: 10px;
  background: unset !important;
  border: 2px dashed #a5c8d7;
  width: 212px;
  height: 206px;
  position: relative;
  border-radius: 5px;
  cursor: pointer;
`;
const StyledImageBackground = styled.div`
  background-color: #f1f8f8;
  text-align: center;
`;

const StyledImageSpace = styled.div`
  padding: 1em;
`;
