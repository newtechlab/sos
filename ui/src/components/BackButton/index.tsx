import { Button } from "semantic-ui-react";

interface BackButtonProps {
    goBack: () => void
}

export default function BackButton(props: BackButtonProps) {
    const { goBack } = props;

 return <Button basic circular color="teal" onClick={() => {
    goBack();
        }}>Tilbake</Button>
}


