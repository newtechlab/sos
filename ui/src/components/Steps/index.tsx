import _ from "lodash";
import { Icon, Step } from "semantic-ui-react";

export interface StepsState {
    activeStepId:  number;
    completedGroups: Set<StepGroupType>;
    stepGroups: Map<StepGroupType, StepGroup>;
    steps: Array<StepDefinition>;
}

export enum StepGroupType {
    FAMILY,
    MONEY_IN,
    MONEY_OUT,
    RESULTS
}

export interface StepGroup {
    title: string;
    description: string;
}

export interface StepDefinition {
    id: number;
    group: StepGroupType;
    completed: boolean;
    path: string;
    heading: string;
    description?: string;
}

export interface StepsProps {
  steps: StepsState
}

export default function Steps(props: StepsProps) {
    const { steps } = props
    const ActiveIcon = "arrow alternate circle down";
    const CompleteIcon = "check circle";
    const stepsArray = Array.from( steps.stepGroups )
    return <Step.Group widths={4}>
            { stepsArray.map(([id, step]) => {
                const isActive = steps.activeStepId === id;
                return <Step key={id} active={isActive}>
                    {  isActive && <Icon name={ActiveIcon} color="teal" /> }
                    {  steps.completedGroups.has(id) && <Icon name={CompleteIcon} color="green" /> } 
                    <Step.Content>
                    <Step.Title>{ steps.stepGroups.get(id)?.title }</Step.Title>
                    <Step.Description> { steps.stepGroups.get(id)?.description }</Step.Description>
                    </Step.Content>
                </Step>
            })}
        </Step.Group>
}