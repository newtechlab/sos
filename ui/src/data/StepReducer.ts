import { StepDefinition } from "../components/Steps";

export const stepReducer = (accumulation: Array<StepDefinition>, current: StepDefinition): Array<StepDefinition> => {
    const prevItem: StepDefinition | undefined = (accumulation && accumulation.length > 0) ? accumulation[accumulation.length-1] : undefined
    const prevItemCompleted = prevItem ? prevItem.completed : false;
    const thisIsCompleted = current.active || current.completed;
    const newStep: StepDefinition = {
      ...current,
      completed: thisIsCompleted,
      active: prevItemCompleted && !thisIsCompleted
    } 
    return accumulation.concat(newStep);
}