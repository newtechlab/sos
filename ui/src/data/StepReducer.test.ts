import { stepReducer } from './StepReducer';
import { StepDefinition } from '../components/Steps';

test('step reducer to work with one step', () => {
  const current: StepDefinition = {
    id: 0,
    active: false,
    completed: false,
    title: "title",
    description: "description",
    path: "foo"
  }
  const accumulation: Array<StepDefinition> = [];
  const foo = stepReducer(accumulation, current)
  expect(foo).toMatchSnapshot();
});

test('step reducer to mark next step as complete', () => {
  const first: StepDefinition = {
    id: 0,
    active: true,
    completed: false,
    title: "title 0",
    description: "description 0",
    path: "foo 0"
  }
  const second: StepDefinition = {
    id: 1,
    active: false,
    completed: false,
    title: "title 1",
    description: "description 1",
    path: "foo 1"
  };
  const stepOne = stepReducer([], first)
  const foo = stepReducer(stepOne, second)
  expect(foo).toMatchSnapshot();
});
