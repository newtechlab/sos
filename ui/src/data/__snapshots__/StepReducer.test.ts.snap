// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`step reducer to mark next step as complete 1`] = `
Array [
  Object {
    "active": false,
    "completed": true,
    "description": "description 0",
    "id": 0,
    "path": "foo 0",
    "title": "title 0",
  },
  Object {
    "active": true,
    "completed": false,
    "description": "description 1",
    "id": 1,
    "path": "foo 1",
    "title": "title 1",
  },
]
`;

exports[`step reducer to work with one step 1`] = `
Array [
  Object {
    "active": false,
    "completed": false,
    "description": "description",
    "id": 0,
    "path": "foo",
    "title": "title",
  },
]
`;
